### PR TITLE
implement SPLIT Aql-Function in C++

### PR DESCRIPTION
**Pull request by dothebart**

### CODE REVIEW

File: arangod/Aql/Functions.cpp [hunk 1]
```diff
@@ -1510,6 +1544,141 @@ AqlValue Functions::Like(arangodb::aql::Query* query,
...
+
+    // these are edge cases which are documented to have these return values:
+    if (limitNumber < 0) {
+      return AqlValue(AqlValueHintNull());
+    }
+    if (limitNumber == 0) {
+      VPackBuilder result;
+      result.openArray();
+      result.close();
+      return AqlValue(result);
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: replace with `return AqlValue(VPackSlice::emptyArraySlice());`?

File: arangod/Aql/Functions.cpp [hunk 2]
```diff
@@ -1510,6 +1544,141 @@ AqlValue Functions::Like(arangodb::aql::Query* query,
...
+  static const uint16_t nrResults = 16;
+  UnicodeString uResults[nrResults];
+  int64_t totalCount = 0;
+  while (true) {
+    UErrorCode errorCode = U_ZERO_ERROR;
+    auto uCount = matcher->split(valueToSplit, uResults, nrResults, errorCode);
+    uint16_t copyThisTime = uCount;
+    
+    if (U_FAILURE(errorCode)) {
+      RegisterWarning(query, "LIKE", TRI_ERROR_QUERY_INVALID_REGEX);
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: should be "SPLIT" here, not "LIKE".

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -1510,6 +1544,141 @@ AqlValue Functions::Like(arangodb::aql::Query* query,
...
+      return AqlValue(AqlValueHintNull());
+    }
+    
+    if ((copyThisTime > 0) && (copyThisTime > nrResults)) {
+      // last hit is the remaining string to be feed into split in a subsequent invocation
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: to be fed?

File: arangod/Aql/Functions.cpp [hunk 4]
```diff
@@ -1510,6 +1544,141 @@ AqlValue Functions::Like(arangodb::aql::Query* query,
...
+  }
+
+  std::string utf8;
+  static const uint16_t nrResults = 16;
+  UnicodeString uResults[nrResults];
+  int64_t totalCount = 0;
+  while (true) {
+    UErrorCode errorCode = U_ZERO_ERROR;
+    auto uCount = matcher->split(valueToSplit, uResults, nrResults, errorCode);
+    uint16_t copyThisTime = uCount;
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: can't `uCount` be negative here? if so, what value will `copyThisTime` have then?

File: arangod/Aql/RegexCache.cpp [hunk 5]
```diff
@@ -46,6 +51,40 @@ icu::RegexMatcher* RegexCache::buildLikeMatcher(char const* ptr, size_t length,
...
+      if (copy.length() == 0) {
+        // one empty string rules them all
+        isEmptyExpression = true;
+        rx = "";
+        break;
+      }
+      if (rx.size() != 0) {
+        rx += '|';
+      }
+      rx += copy;
```
Comment (jsteemann) [inline arangod/Aql/RegexCache.cpp]: shouldn't `copy` be properly escaped for regex usage here?

Comment (dothebart) [reply arangod/Aql/RegexCache.cpp]: used https://gist.github.com/edhemphill/1731633  for testing.

**Pull request merged**